aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -66,14 +66,43 @@ Tensor& asin_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(
 Tensor asin(const Tensor& self) { return unary_op_impl(self, at::asin_out); }
 Tensor& asin_(Tensor& self) { return unary_op_impl_(self, at::asin_out); }
 
-Tensor& abs_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, abs_stub); }
+Tensor& abs_out(Tensor& result, const Tensor& self) {
+  // Mimics abs returning a floating point tensor for complex inputs by default
+  // Note: This is done by running the operation as usual, then checking
+  // for a valid cast from float, not complex, and copying the operation's
+  // result to the expected result type.
+  if (self.is_complex() && !result.is_complex()) {
```

Comment:
I think we should add:
```
if (self.is_complex()) {
   TORCH_CHECK(ztype<c10::typeMetaToScalarType(self.dtype())>::value_t == 
     c10::typeMetaToScalarType(result.dtype()));
}
```
i.e. we should throw an error when self is complex and (self, result) scalar_type pair is not (ComplexFloat, Float), (ComplexDouble, Double) and disallow result type to be complex

Reply:
I see you are using cancast below and that from my understanding additionally allows this case: ```(ComplexFloat, Double)```, then we could alternately just add TORCH_CHECK(!result.is_complex()) above

Reply:
I'm not sure I follow. 'canCast' does not complex -> noncomplex casts (see https://github.com/pytorch/pytorch/blob/c070e8fb26e276da2148b7604b9a93d85740f5e2/c10/core/ScalarType.h#L380) and float, double, and complex types all seem OK for the out type?

Reply:
yeah I meant that our out dtype for ComplexFloat input is float and canCast would promote that to double hence `(ComplexFloat, Double)`